retrieval/targetmanager.go

```diff
@@ -14,139 +14,356 @@
+			break
+		}
+		glog.V(1).Infof("Received update for target group %q", tg.Source)
 
-	targetPool := m.targetPoolForJob(job)
-	targetPool.AddTarget(t)
-	m.poolsByJob[job.GetName()] = targetPool
+		if err := tm.updateTargetGroup(tg, cfg); err != nil {
+			glog.Errorf("Error updating targets: %s", err)
+		}
+	}
 }
 
-func (m *targetManager) ReplaceTargets(job config.JobConfig, newTargets []Target) {
-	m.Lock()
-	defer m.Unlock()
+// fullSource prepends the source by the unique job name.
+//
+// Thus, oscilliating label sets for targets with the same source,
+// but providers from different configs, are prevented.
+func fullSource(cfg config.JobConfig, src string) string {
+	return cfg.GetName() + ":" + src
+}
 
-	targetPool := m.targetPoolForJob(job)
-	targetPool.ReplaceTargets(newTargets)
+// Stop all background processing.
+func (tm *TargetManager) Stop() {
+	tm.stop(true)
 }
 
-func (m *targetManager) Remove(t Target) {
-	panic("not implemented")
+// stop background processing of the target manager. If removeTargets is true,
+// existing targets will be stopped and removed.
+func (tm *TargetManager) stop(removeTargets bool) {
+	tm.m.Lock()
+	defer tm.m.Unlock()
+
+	if !tm.running {
+		return
+	}
+
+	glog.Info("Stopping target manager...")
+	defer glog.Info("Target manager stopped.")
+
+	for _, provs := range tm.providers {
+		for _, p := range provs {
+			p.Stop()
+		}
+	}
+
+	if removeTargets {
+		tm.removeTargets(nil)
+	}
+
+	tm.running = false
 }
 
-func (m *targetManager) AddTargetsFromConfig(config config.Config) {
-	for _, job := range config.Jobs() {
-		if job.SdName != nil {
-			m.Lock()
-			m.targetPoolForJob(job)
-			m.Unlock()
+// removeTargets stops and removes targets for sources where f(source) is true
+// or if f is nil. This method is not thread-safe
+func (tm *TargetManager) removeTargets(f func(string) bool) {
+	if f == nil {
+		f = func(string) bool { return true }
+	}
+	var wg sync.WaitGroup
+	for src, targets := range tm.targets {
+		if !f(src) {
 			continue
 		}
+		wg.Add(len(targets))
+		for _, target := range targets {
+			go func(t Target) {
+				t.StopScraper()
+				wg.Done()
+			}(target)
+		}
+		delete(tm.targets, src)
+	}
+	wg.Wait()
+}
+
+// updateTargetGroup creates new targets for the group and replaces the old targets
+// for the source ID.
+func (tm *TargetManager) updateTargetGroup(tgroup *config.TargetGroup, cfg config.JobConfig) error {
+	newTargets, err := tm.targetsFromGroup(tgroup, cfg)
+	if err != nil {
+		return err
+	}
+	src := fullSource(cfg, tgroup.Source)
 
-		for _, targetGroup := range job.TargetGroup {
-			baseLabels := clientmodel.LabelSet{
-				clientmodel.JobLabel: clientmodel.LabelValue(job.GetName()),
+	tm.m.Lock()
+	defer tm.m.Unlock()
+
+	oldTargets, ok := tm.targets[src]
+	if ok {
+		// Replace the old targets with the new ones while keeping the state
+		// of intersecting targets.
+		for i, tnew := range newTargets {
+			var match Target
+			for i, told := range oldTargets {
```

Comment:
You're reusing (and shadowing) `i` here, which is already used in the outer loop. Please don't :)